lib/metasploit/framework/login_scanner/base.rb

```diff
@@ -202,7 +228,33 @@ def scan!
             ignored_users = Set.new
             first_attempt = true
 
-            each_credential do |credential|
+            # previous password so we know when a new cycle begins
+            prev_password = nil
+            cycles = 0
+
+            # run through each_credential and put yielded results into array
+            all_creds = []
+            each_credential { |c| all_creds << c }
```

Comment:
This builds a potentially quite large array in memory. Instead, I think it would make more sense to add a `password_spray` option to `lib/metasploit/framework/credential_collection.rb` and have it generate the creds password-first.

Reply:
This doesn't really make sense to me. Even with a thousand usernames and passwords, it seems unlikely that the (temporary) memory consumption would be an issue, and 99% of the time people aren't using lists even close to that size.

Reply:
```
idle msfconsole:
total kB          753196  210092  199732

100 * 100 = 10k creds:
total kB          788244  210632  199936

1000 * 1000 = 1mil creds:
total kB         1061540  428364  417760
```

Reply:
I'm not sure how to do that since the creds are yielded one at a time, hence this approach of collecting them into an array and sorting it. Maybe I don't understand what you mean.

And are we sure this is ever going to be a problem? Even with a million Credential objects, it's not so bad:

```
idle msfconsole:
total kB          753196  210092  199732

100 * 100 = 10k creds:
total kB          788244  210632  199936

1000 * 1000 = 1mil creds:
total kB         1061540  428364  417760
```

Reply:
Yes, these lists can become quite large. If you had a medium-large number of users, say 35k, and 100 common passwords/variants for each, that would be more than an extra gigabyte of memory according to you tests, which is more than enough to put a noticeable strain on the Ruby VM garbage collections, especially over the amount of time it would take to attempt all those logins. For a pentester on a short gig that might be a stretch, but this code is also used by Pro to do things of that scale and larger.

Since pairs are generated one by one you need to think of this less than a sort of generated credentials, to changing the order in which they are generated. If you look at `#each` on the credential collection (https://github.com/rapid7/metasploit-framework/blob/master/lib/metasploit/framework/credential_collection.rb#L112), you can see that it goes through all the usernames one by one and yields all the passwords that it can find for each one. To generate a spray, then, you would need to go through each of the passwords and then yield all the usernames that are part of the collection. This is similar to row-major vs column-major if you have worked with matrices: https://en.wikipedia.org/wiki/Row-_and_column-major_order

Reply:
I do understand what you're saying, but I still can't imagine anyone would attempt 3.5 million guesses over the Internet. That would take months to finish.

Anyway, I've lost interest here. Thanks for taking the time though!

Reply:
I can't see someone really using more than a couple of passwords just to avoid account lockout. You know, Summer2018 across all usernames, that sort of thing. Think I concur with @rwhitcroft here, technically this can blow up, but practically you'd never get there fast enough for it to matter much. We could 'msf5' tag this if pro blowup is a concern.